File: lint-arwaky/archive/capabilities/data_flow_analyzer.py
"""DataFlowAnalyzer — Capability for analyzing data flow patterns."""

from __future__ import annotations
from ..taxonomy import DataFlowList, FilePath, LineNumber, SymbolName

import logging
import re


from ..contract import IDataFlowProtocol


from ..contract import IFileSystemPort

from ..contract import IScopeBoundaryProtocol
from .scope_boundary_analyzer import ScopeBoundaryAnalyzer

logger = logging.getLogger(__name__)


class DataFlowAnalyzer(IDataFlowProtocol):
    """Capability for tracking variable lifecycle in JS/TS files."""

    def __init__(
        self, fs: IFileSystemPort, scope: IScopeBoundaryProtocol | None = None
    ):
        self._fs = fs
        self._scope = scope or ScopeBoundaryAnalyzer(fs)

    def find_flow(
        self, file_path: FilePath, var_name: SymbolName, start_line: LineNumber
    ) -> DataFlowList:
        flows = self._find_flow_raw(
            self._fs, file_path, var_name, int(start_line), self._scope
        )
        return DataFlowList(values=flows)

    def _try_build_entry(
        self,
        raw_line: str,
        line_no: int,
        word_pattern: re.Pattern,
        assign_pattern: re.Pattern,
        mutation_pattern: re.Pattern,
        seen: set[str],
    ) -> str | None:
        """Return a formatted flow entry for this line, or None if not relevant."""
        if not word_pattern.search(raw_line):
            return None
        stripped = raw_line.strip()
        match = mutation_pattern.search(raw_line)
        if match:
            entry = f"Line {line_no} [Mutation '{match.group(1)}']: {stripped}"
        elif assign_pattern.search(raw_line):
            entry = f"Line {line_no} [Assignment]: {stripped}"
        elif word_pattern.search(raw_line):
            entry = f"Line {line_no} [Usage]: {stripped}"
        else:
            entry = None
        if entry and entry in seen:
            return None
        if entry:
            seen.add(entry)
        return entry

    def _walk_calls_in_body(
        self,
        lines: list[str],
        word_pattern: re.Pattern,
        assign_pattern: re.Pattern,
        mutation_pattern: re.Pattern,
        scope_start: int | None,
        scope_end: int | None,
    ) -> list[str]:
        """Walk through lines dispatching each to pattern matchers and collecting flow entries."""
        flows: list[str] = []
        seen: set[str] = set()
        for i, raw_line in enumerate(lines):
            line_no = i + 1
            if scope_start is not None and line_no < scope_start:
                continue
            if scope_end is not None and line_no > scope_end:
                break
            entry = self._try_build_entry(
                raw_line,
                line_no,
                word_pattern,
                assign_pattern,
                mutation_pattern,
                seen,
            )
            if entry:
                flows.append(entry)
        return flows

    def _find_flow_raw(
        self,
        fs: IFileSystemPort,
        file_path: FilePath,
        var_name: SymbolName,
        start_line: int | None = None,
        scope_analyzer: IScopeBoundaryProtocol | None = None,
    ) -> list[str]:
        """Track assignments and usages of a variable in a JS/TS file."""
        if not fs.exists(file_path):
            return []

        try:
            content = fs.read_text(file_path)
            lines = content.value.splitlines()
        except Exception:
            return []

        vn = str(var_name)
        scope_start = None
        scope_end = None
        if scope_analyzer:
            start_line_vo = LineNumber(value=start_line) if start_line else None
            s_start_vo, s_end_vo = scope_analyzer.find_scope_bounds(
                content, start_line_vo
            )
            scope_start = int(s_start_vo) if s_start_vo is not None else None
            scope_end = int(s_end_vo) if s_end_vo is not None else None

        word_pattern = re.compile(rf"\b{re.escape(vn)}\b")
        assign_pattern = re.compile(
            rf"(?:const|let|var)\s+{re.escape(vn)}\s*=|(?<![=!<>]){re.escape(vn)}\s*="
        )
        mutation_pattern = re.compile(
            rf"\b{re.escape(vn)}\.(push|pop|shift|unshift|splice|sort|reverse|"
            rf"set|delete|add|assign|merge|update|append|extend)\b"
        )

        return self._walk_calls_in_body(
            lines,
            word_pattern,
            assign_pattern,
            mutation_pattern,
            scope_start,
            scope_end,
        )

File: lint-arwaky/src-rust/capabilities/data_flow_analyzer.rs
// data_flow_analyzer — Capability for analyzing data flow patterns.
// Implements IDataFlowProtocol: find_flow — tracks variable lifecycle in JS/TS files.

use std::fs;
use std::collections::HashSet;
use regex::Regex;
use crate::taxonomy::{FilePath, LineNumber, SymbolName};
use crate::capabilities::scope_boundary_analyzer::ScopeBoundaryAnalyzer;

/// A single data flow entry describing a variable's usage at a line.
pub struct DataFlowEntry {
    pub line: usize,
    pub kind: String, // "Assignment", "Mutation", "Usage"
    pub content: String,
}

/// Business logic for tracking variable lifecycle in JS/TS files.
pub struct DataFlowAnalyzer {
    scope: ScopeBoundaryAnalyzer,
}

impl DataFlowAnalyzer {
    pub fn new() -> Self {
        Self {
            scope: ScopeBoundaryAnalyzer::new(),
        }
    }

    /// Track assignments and usages of a variable in a JS/TS file.
    pub fn find_flow(
        &self,
        file_path: &str,
        var_name: &str,
        start_line: Option<usize>,
    ) -> Vec<DataFlowEntry> {
        let Ok(content) = fs::read_to_string(file_path) else {
            return vec![];
        };

        // Determine scope bounds
        let (scope_start, scope_end) = if let Some(line) = start_line {
            self.scope.find_scope_bounds(&content, Some(line))
        } else {
            (None, None)
        };

        let word_pattern = Regex::new(&format!(r"\b{}\b", regex::escape(var_name))).unwrap();
        let assign_pattern = Regex::new(&format!(
            r"(?:const|let|var)\s+{}\s*=|(?<![=!<>]){}s*=",
            regex::escape(var_name), regex::escape(var_name)
        )).unwrap();
        let mutation_pattern = Regex::new(&format!(
            r"\b{}\.(?:push|pop|shift|unshift|splice|sort|reverse|set|delete|add|assign|merge|update|append|extend)\b",
            regex::escape(var_name)
        )).unwrap();

        let mut flows: Vec<DataFlowEntry> = Vec::new();
        let mut seen: HashSet<String> = HashSet::new();

        for (i, line_str) in content.lines().enumerate() {
            let line_no = i + 1;

            if let Some(s) = scope_start {
                if line_no < s { continue; }
            }
            if let Some(e) = scope_end {
                if line_no > e { break; }
            }

            if !word_pattern.is_match(line_str) {
                continue;
            }

            let stripped = line_str.trim();
            let entry_str = if let Some(m) = mutation_pattern.find(line_str) {
                // Extract the method name from the match
                let method = m.as_str().split('.').nth(1).unwrap_or("mutation");
                format!("Line {} [Mutation '{}']: {}", line_no, method, stripped)
            } else if assign_pattern.is_match(line_str) {
                format!("Line {} [Assignment]: {}", line_no, stripped)
            } else {
                format!("Line {} [Usage]: {}", line_no, stripped)
            };

            if seen.contains(&entry_str) {
                continue;
            }
            seen.insert(entry_str.clone());

            let kind = if mutation_pattern.is_match(line_str) {
                "Mutation".to_string()
            } else if assign_pattern.is_match(line_str) {
                "Assignment".to_string()
            } else {
                "Usage".to_string()
            };

            flows.push(DataFlowEntry {
                line: line_no,
                kind,
                content: stripped.to_string(),
            });
        }

        flows
    }
}
